File: Z-analytic/Z2.py
from scipy.special import dawsn  # requires numba-scipy
import numpy as np
import numba as nb
import Z1

# ======================================== BASE ========================================
SIGMAS = np.asarray([-1, 1])


@nb.njit
def F2(x):
    return dawsn(x)


@nb.njit
def _Z2_E_terms_norm(h, l):
    def func(s1, s2): return np.exp((s1+s2)*h+s1*s2*l-2*h-l)
    return np.asarray([[func(s1, s2) for s2 in SIGMAS] for s1 in SIGMAS])


@nb.njit
def _Z2_F_terms(h, l):
    def func(s1, s2): return F2((h+(s1+s2)*l)/np.sqrt(2*l))
    return np.asarray([[func(s1, s2) for s2 in SIGMAS] for s1 in SIGMAS])


@nb.njit
def _Z2_sum_terms(terms, sigmas_func, *sigmas_args):
    result = 0
    for s1 in range(2):
        for s2 in range(2):
            result += sigmas_func(SIGMAS[s1], SIGMAS[s2], *sigmas_args)*terms[s1, s2]
    return result


@nb.njit
def cF2Function_norm(h, l):
    _terms = _Z2_E_terms_norm(h, l)*_Z2_F_terms(h, l)
    def _sigmas_func_12(s1, s2, *sigmas_args): return s1*s2
    return _Z2_sum_terms(_terms, _sigmas_func_12, 1, 1)


@nb.njit
def EFunction_norm(h, l, arg_s1, arg_s2):
    def _sigmas_func(s1, s2, arg_s1, arg_s2): return (s1 if arg_s1 else 1)*(s2 if arg_s2 else 1)
    return _Z2_sum_terms(_Z2_E_terms_norm(h, l), _sigmas_func, arg_s1, arg_s2)


# ======================================== MOMENTS ========================================
M_SQRT2 = np.sqrt(2)


def calc_Z2_norm(h, l):
    return M_SQRT2*(2*np.pi)**2/(h*np.sqrt(l))*cF2Function_norm(h, l)


def calc_m(h, l):
    if h == 0:
        return 0
    if l == 0:
        return Z1.calc_m_single(h)
    return -1/(2*h) - h/(2*l) + M_SQRT2*EFunction_norm(h, l, 1, 1)/(4*np.sqrt(l)*cF2Function_norm(h, l))


def calc_eta(h, l):
    if h == 0:
        return Z1.calc_m_single(l)
    if l == 0:
        return Z1.calc_m_single(h)**2
    return -1 - 1/(2*l) + h**2/(2*l**2) + M_SQRT2*EFunction_norm(h, l, 1, 0)/(2*np.sqrt(l)*cF2Function_norm(h, l)) - M_SQRT2*h*EFunction_norm(h, l, 1, 1)/(4*l**(3/2)*cF2Function_norm(h, l))


def calc_upsilon(m, eta, h, l):
    if h == 0 and l == 0:
        return 1/3
    if h == 0:
        return eta/(l*(eta+1))
    if l == 0:
        return m/h
    m_eta = -eta*(l + h**2)/(2*l*h) - m/l + (2*l*m*h + l + h**2)*EFunction_norm(h, l, 0, 0)/(2*l*h*EFunction_norm(h, l, 1, 1)) - 1/(2*l*h)
    return .5*(1-m_eta/m)


def calc_eta2(m, eta, h, l):
    if h == 0 and l == 0:
        return 1/3
    if h == 0:
        return 1-2*eta/l
    if l == 0:
        m_h = m/h
        return 6*m_h*m_h - 4*m_h + 1
    return eta*(-3*l + h**2)/(2*l**2) + m*h*(l + 1)/l**2 + (2*l**2 + 1)/(2*l**2) - (2*l*m*h + l + h**2)*EFunction_norm(h, l, 0, 0)/(2*l**2*EFunction_norm(h, l, 1, 1))


_, calc_Z2_norm = Z1.single_and_vectorize(calc_Z2_norm)
_, calc_m = Z1.single_and_vectorize(calc_m)
_, calc_eta = Z1.single_and_vectorize(calc_eta)
calc_upsilon_single, calc_upsilon = Z1.single_and_vectorize(calc_upsilon)
calc_eta2_single, calc_eta2 = Z1.single_and_vectorize(calc_eta2)


def calc_m_par_2(m, eta, h, l):
    if h == 0:
        m_h = (1+eta)/3
        return 1 - 2*m_h + 2*m_h*eta/(eta+1)
    if l == 0:
        return 1 - 2*m/h
    upsilon = calc_upsilon_single(m, eta, h, l)
    return 1 - 2*m/h + 2*l*m/h*upsilon


def calc_psi0(m, eta, h, l):
    upsilon = calc_upsilon_single(m, eta, h, l)
    eta2 = calc_eta2_single(m, eta, h, l)
    if h == 0:
        m_h = (1+eta)/3
    else:
        m_h = m/h
    return 1.5*m_h*upsilon + .25*(eta2-1)


_, calc_m_par_2 = Z1.single_and_vectorize(calc_m_par_2)
_, calc_psi0 = Z1.single_and_vectorize(calc_psi0)


# ======================================== FINDING ========================================
@nb.njit
def m_upsilon_approx(m, eta):
    m2 = m * m
    zeta = (eta - m2) / (1 - m2)
    zeta2 = zeta * zeta

    mu_p = Z1.mu_p_approx(m)
    m_upsilon = (1 - eta) / (1 - m2) * m * mu_p + 0.0467246 * m * (1 - m) * zeta * (1 - zeta) * (1 - 7.96819 * m - 0.939775 * zeta + 14.0242 * m * zeta + 20.9323 * m2 + 7.65054 * zeta2 - 14.5937 * m2 * zeta - 5.35767 * m * zeta2 - 13.542 * m * m2 - 3.72091 * zeta * zeta2) + (0.00349416 + (0.0191611 + (-0.0367299 + (-0.0684992 + (0.149972 - 0.067404 * m2) * m2) * m2) * m2) * m2) * m
    return m_upsilon


@nb.njit
def rho_approx(m, eta):
    m2 = m * m
    zeta = (eta - m2) / (1 - m2)
    rho = 1 - .5 * zeta - .514 * (1 - zeta) * zeta * (1 - (.232 + (1.681 - 1.466 * m) * m) * m + (-.123 * m - 1.041 + (0.858 - 0.356 * zeta) * zeta) * zeta)
    return rho


@nb.njit
def coeffs_approx(m, eta):
    mu_p = Z1.mu_p_approx(m)
    upsilon = m_upsilon_approx(m, eta) / m
    rho = rho_approx(m, eta)

    return rho * m / mu_p, (1 - rho) / upsilon


@nb.njit
def find_coeffs_single(m, eta, eps=1e-8):
    if eta < m * m:
        return np.nan, np.nan

    if m == 1 or eta == 1:
        return np.nan, np.nan
    
    if eta == m * m:
        return Z1.find_coeffs_single(m), 0

    if m == 0 and eta == 0:
        return 0, 0
    
    if m == 0:
        return 0, Z1.find_coeffs_single(eta)


    h, l = coeffs_approx(m, eta)

    delta_h, delta_l = 1, 1
    steps = 0
    while delta_h * delta_h + delta_l * delta_l > eps * eps and steps < 64:
        if h == 0 or l == 0:
            break
        if l < 0:
            break

        _l = 1 / l
        _h = 1 / h
        _sqrt_l = np.sqrt(_l)
        _exp_2l = np.exp(-2 * l)
        _exp_2h = np.exp(-2 * h)
        cF2     = _exp_2h * _exp_2h * F2((.5) * M_SQRT2 * _sqrt_l * (h - 2 * l)) - 2 * _exp_2h * _exp_2l * F2((.5) * M_SQRT2 * _sqrt_l * h) + F2((.5) * M_SQRT2 * _sqrt_l * (h + 2 * l))
        _cF2 = 1 / cF2
        E12_cF2 = _cF2 * (_exp_2h * _exp_2h - 2 * _exp_2h * _exp_2l + 1)
        E1_cF2 = _cF2 * (1 - _exp_2h * _exp_2h)
        E_cF2 = _cF2 * (_exp_2h * _exp_2h + 2 * _exp_2h * _exp_2l + 1)

        expr_m = (.25) * M_SQRT2 * E12_cF2 * _sqrt_l - .5 * _h - .5 * _l * h
        expr_m_diff_h = (.25) * M_SQRT2 * E12_cF2 * _l * _sqrt_l * h + (.5) * M_SQRT2 * E1_cF2 * _sqrt_l + (.5) * _h * _h + _l * (-.25 * E12_cF2 * E12_cF2 - 1.0 / 2.0)
        expr_m_diff_l = -.25 * E12_cF2 * E1_cF2 * _l - 0.125 * M_SQRT2 * E12_cF2 * _l * _l * _sqrt_l * h * h - 0.125 * M_SQRT2 * E12_cF2 * _l * _sqrt_l + _l * _l * ((0.125) * E12_cF2 * E12_cF2 * h + (.5) * h) + _sqrt_l * ((.25) * M_SQRT2 * E12_cF2 + (.25) * M_SQRT2 * E_cF2)

        expr_eta = -.25 * M_SQRT2 * E12_cF2 * _l * _sqrt_l * h + (.5) * M_SQRT2 * E1_cF2 * _sqrt_l + (.5) * _l * _l * h * h - .5 * _l - 1
        expr_eta_diff_h = -.5 * E12_cF2 * E1_cF2 * _l - .25 * M_SQRT2 * E12_cF2 * _l * _l * _sqrt_l * h * h - .25 * M_SQRT2 * E12_cF2 * _l * _sqrt_l + _l * _l * ((.25) * E12_cF2 * E12_cF2 * h + h) + _sqrt_l * ((.5) * M_SQRT2 * E12_cF2 + (.5) * M_SQRT2 * E_cF2)
        expr_eta_diff_l = (0.125) * M_SQRT2 * E12_cF2 * _l * _l * _l * _sqrt_l * h * h * h - .5 * E1_cF2 * E1_cF2 * _l + M_SQRT2 * E1_cF2 * _sqrt_l + _l * _l * _l * (-0.125 * E12_cF2 * E12_cF2 * h * h - h * h) + _l * _l * _sqrt_l * ((0.375) * M_SQRT2 * E12_cF2 * h - .25 * M_SQRT2 * E1_cF2 * h * h) + _l * _l * ((.5) * E12_cF2 * E1_cF2 * h + .5) + _l * _sqrt_l * (-.25 * M_SQRT2 * E12_cF2 * h - .25 * M_SQRT2 * E1_cF2 - .25 * M_SQRT2 * E_cF2 * h)

        D = expr_m_diff_h * expr_eta_diff_l - expr_m_diff_l * expr_eta_diff_h
        if D == 0:
            break
        _D = 1 / D
        delta_h = ((expr_m - m) * expr_eta_diff_l - (expr_eta - eta) * expr_m_diff_l) * _D
        delta_l = ((expr_eta - eta) * expr_m_diff_h - (expr_m - m) * expr_eta_diff_h) * _D

        h -= delta_h
        l -= delta_l
        steps += 1

    if delta_h * delta_h + delta_l * delta_l > eps * eps:
        return np.nan, np.nan

    return h, l


@nb.njit(parallel=True)
def find_coeffs_vectorize(m, eta, eps=1e-8):
    h = np.empty_like(m)
    l = np.empty_like(m)

    for i in nb.prange(m.size):
        result = find_coeffs_single(m.flat[i], eta.flat[i], eps)
        h.flat[i] = result[0]
        l.flat[i] = result[1]

    return h, l


def find_coeffs(m, eta, eps=1e-8):
    if isinstance(m, np.ndarray):
        return find_coeffs_vectorize(m, eta, eps)
    else:
        return find_coeffs_single(m, eta, eps)

File: src/Z2.cpp
#include <math.h>

#include "../include/Z2.hpp"
#include "../include/special.hpp"

// #include <iostream>

const int SIGMAS[2] = { -1, 1 };

// =======================================================================================

double cF2_norm_symmetrical(double h, double l) {
    double result = 0;
    for (int is1 = 0; is1 < 2; is1++)
        for (int is2 = 0; is2 < 2; is2++) {
            int s1 = SIGMAS[is1];
            int s2 = SIGMAS[is2];
            result += s1 * s2 * exp((s1 + s2 - 2) * h + (s1 * s2 - 1) * l) * dawson((h + (s1 + s2) * l) / sqrt(2 * l));
        }

    return result;
}

double cF2_symmetrical(double h, double l) {
    return cF2_norm_symmetrical(h, l) * exp(2 * h + l);
}

double E_norm_symmetrical(double h, double l, bool arg_s1, bool arg_s2) {
    double result = 0;
    for (int is1 = 0; is1 < 2; is1++)
        for (int is2 = 0; is2 < 2; is2++) {
            int s1 = SIGMAS[is1];
            int s2 = SIGMAS[is2];

            int factor = 1;
            factor *= arg_s1 ? s1 : 1;
            factor *= arg_s2 ? s2 : 1;

            result += factor * exp((s1 + s2 - 2) * h + (s1 * s2 - 1) * l);
        }

    return result;
}

double E_symmetrical(double h, double l, bool arg_s1, bool arg_s2) {
    return E_norm_symmetrical(h, l, arg_s1, arg_s2) * exp(2 * h + l);
}

// =======================================================================================

void Z2_symmetrical::calc_base() {
    // _h = fabs(h) > 1e-4 ? 1 / h : NAN;
    // _l = l > 1e-4 ? 1 / l : NAN;
    _h = 1 / h;
    _l = 1 / l;

    sqrt_l  = sqrt(l);
    _sqrt_l = 1 / sqrt_l;
    // _sqrt_l = sqrt_l > 1e-4 ? 1 / sqrt_l : NAN;

    _exp_2l = exp(-2 * l);
    _exp_2h = exp(-2 * h);

    m_h = NAN;
    // if (!std::isnan(_h) || !std::isnan(_l)) {
    cF2_norm = _exp_2h * _exp_2h * dawson((.5) * M_SQRT2 * _sqrt_l * (h - 2 * l)) - 2 * _exp_2h * _exp_2l * dawson((.5) * M_SQRT2 * _sqrt_l * h) + dawson((.5) * M_SQRT2 * _sqrt_l * (h + 2 * l));

    E12_norm = _exp_2h * _exp_2h - 2 * _exp_2h * _exp_2l + 1;
    E1_norm  = 1 - _exp_2h * _exp_2h;
    E_norm   = _exp_2h * _exp_2h + 2 * _exp_2h * _exp_2l + 1;
    // } else sqrt_l = _sqrt_l = cF2_norm = E12_norm = E1_norm = E_norm = NAN;

    // m_llb    = fabs(h) > 1e-4 ? (1 + _exp_2h) / (1 - _exp_2h) - _h : h / 3;
    // m_h_llb  = fabs(h) > 1e-4 ? m_llb * _h : 1. / 3.;
    // eta_prmg = l > 1e-4 ? (1 + _exp_2l) / (1 - _exp_2l) - _l : l / 3;
}

void Z2_symmetrical::calc_m() {
    // if (fabs(h) > 1e-2 && l > 1e-1) {
    m   = -.25 * (-M_SQRT2 * E12_norm * h * l + 2 * cF2_norm * h * h * sqrt_l + 2 * cF2_norm * l * sqrt_l) / (cF2_norm * h * l * sqrt_l);
    m_h = m * _h;
    // } else if (l <= 1e-1) {
    //     m   = m_llb + l * (m_llb - m_llb * m_llb * m_llb - 2 * m_llb * m_h_llb);
    //     m_h = m_h_llb + l * (m_h_llb - m_h_llb * m_llb * m_llb - 2 * m_h_llb * m_h_llb);
    // } else {  // fabs(h) <= 1e-2
    //     m_h = (1 + eta_prmg) / 3;
    //     m   = h * m_h;
    // }
}

void Z2_symmetrical::calc_eta() {
    // if (fabs(h) > 1e-2 && l > 1e-1)
    eta = .25 * (-M_SQRT2 * E12_norm * h * l * l + 2 * M_SQRT2 * E1_norm * l * l * l + 2 * cF2_norm * h * h * l * sqrt_l - 4 * cF2_norm * l * l * l * sqrt_l - 2 * cF2_norm * l * l * sqrt_l) / (cF2_norm * l * l * l * sqrt_l);
    // else if (l <= 1e-1) eta = m_llb * m_llb + l * (1 - m_llb * m_llb * m_llb * m_llb + 6 * m_h_llb * m_h_llb - 4 * m_h_llb);
    // else eta = eta_prmg;  // fabs(h) <= 1e-2
}

void Z2_symmetrical::calc_upsilon() {
    // if (l <= 1e-2 || (fabs(h) <= 1e-1 && l < .1)) {
    //     double tmp = fabs(h) > 1e-4 ? -9 * m_h_llb * _h * _h + 6 * _h * _h - _h / m_llb : 0;
    //     upsilon    = m_h_llb + l * (-m_llb * m_llb * m_h_llb + m_llb * m_llb / 2 - 4 * m_h_llb * m_h_llb + 3 * m_h_llb - .5 + tmp);
    //     m_eta      = m * (1 - 2 * upsilon);
    // } else if (fabs(h) <= 1e-2 || (fabs(h) <= 1e-1 && l < .5)) {
    //     upsilon = eta * _l / (eta + 1);
    //     m_eta   = m * (1 - 2 * upsilon);
    // } else {
    m_eta   = -.5 * (E12_norm * eta * (h * h) + E12_norm * eta * l + 2 * E12_norm * h * m + E12_norm - E_norm * (h * h) - 2 * E_norm * h * l * m - E_norm * l) / (E12_norm * h * l);
    upsilon = .5 * (1 - m_eta / m);
    // }
}

void Z2_symmetrical::calc_eta2() {
    // if (l > 1e-2)
    eta2 = 1 + 2 * _l * (h * m * upsilon - eta);
    // else if (l <= 1e-2) {
    //     double tmp = fabs(h) > 1e-1 ? 90 * m_h_llb * m_h_llb * _h * _h - 60 * m_h_llb * _h * _h + 10 * _h * _h : 0;
    //     eta2       = 6 * m_h_llb * m_h_llb - 4 * m_h_llb + 1 + l * (-6 * m_llb * m_llb * m_h_llb * m_h_llb + 4 * m_llb * m_llb * m_h_llb + 12 * m_h_llb * m_h_llb - 4 * m_h_llb + tmp);
    // }
}

void Z2_symmetrical::calc_mh2() {
    mh2 = 1 - 2 * m_h + 2 * l * m_h * upsilon;
}

void Z2_symmetrical::calc_psi0() {
    psi0 = 1.5 * m_h * upsilon + .25 * (eta2 - 1);
}

void Z2_symmetrical::calc_Z2_norm() {
    // if (fabs(h) > 1e-2 && l > 1e-2)
    Z2_norm = M_SQRT2 * (2 * M_PI) * (2 * M_PI) * cF2_norm / (h * sqrt_l);
    // else if (l <= 1e-2) {
    //     double Z = 4 * M_PI * (fabs(h) > 1e-4 ? .5 * (1 - _exp_2h) * _h : exp(-h) * (1 + h * h / 6));
    //     Z2_norm  = Z * Z * (1 + l * eta);
    // } else {  // fabs(h) <= 1e-2
    //     double Z = (4 * M_PI) * (4 * M_PI) * (l > 1e-4 ? .5 * (1 - _exp_2l) * _l : exp(-l) * (1 + l * l / 6));
    //     Z2_norm  = Z * (1 + h * m);
    // }
}

// =======================================================================================

void Z2_symmetrical::calc_from_coeffs() {
    calc_base();
    calc_m();
    calc_eta();
    calc_upsilon();
    calc_eta2();
    calc_mh2();
    calc_psi0();
    calc_Z2_norm();
    // if (eta < m * m || eta2 > 1) m = eta = upsilon = eta2 = mh2 = psi0 = NAN;
}

int Z2_symmetrical::calc_from_moments(double h0, double l0, double eps, int max_steps) {
    int steps = 0;

    if (eta < m * m) h = l = NAN;
    else if (m >= 1 || eta >= 1) h = l = NAN;
    else if (eta == m * m) {
        h = invL(m);
        l = 0;
    } else if (m == 0 && eta == 0) h = l = 0;
    else if (m == 0) {
        h = 0;
        l = invL(eta);
    } else {
        h = h0;
        l = l0;

        double m_ref = m, eta_ref = eta;

        double delta_h = 1, delta_l = 1;
        while (delta_h * delta_h + delta_l * delta_l > eps * eps && steps < max_steps) {
            if (h == 0 || l == 0) break;
            if (l < 0) break;

            // std::cout << h << ' ' << l << '\n';

            calc_base();

            calc_m();
            calc_eta();
            calc_upsilon();
            calc_eta2();

            double expr_m_diff_h = eta - 2 * m * m - 2 * m_h + 1;
            double expr_m_diff_l = m_eta - m * eta;

            double expr_eta_diff_h = 2 * expr_m_diff_l;
            double expr_eta_diff_l = eta2 - eta * eta;

            double D = expr_m_diff_h * expr_eta_diff_l - expr_m_diff_l * expr_eta_diff_h;
            if (D == 0) break;
            double _D = 1 / D;
            delta_h   = ((m - m_ref) * expr_eta_diff_l - (eta - eta_ref) * expr_m_diff_l) * _D;
            delta_l   = ((eta - eta_ref) * expr_m_diff_h - (m - m_ref) * expr_eta_diff_h) * _D;

            h -= delta_h;
            l -= delta_l;
            steps += 1;
        }

        if (h != h || l != l) h = l = NAN;
        if (delta_h * delta_h + delta_l * delta_l > eps * eps) h = l = NAN;
    }

    calc_from_coeffs();

    return steps;
}

// =======================================================================================

double cF2_norm_not_symmetrical(double hi, double hj, double l) {
    double result = 0;
    for (int is1 = 0; is1 < 2; is1++)
        for (int is2 = 0; is2 < 2; is2++) {
            int s1 = SIGMAS[is1];
            int s2 = SIGMAS[is2];
            result += s1 * s2 * exp((s1 - 1) * hi + (s2 - 1) * hj + (s1 * s2 - 1) * l) * dawson((hj + s1 * l + s2 * l * hj / hi) / sqrt(2 * l * hj / hi));
        }

    return result;
}

double cF2_not_symmetrical(double hi, double hj, double l) {
    return cF2_norm_not_symmetrical(hi, hj, l) * exp(hi + hj + l);
}

double E_norm_not_symmetrical(double hi, double hj, double l, bool arg_s1, bool arg_s2) {
    double result = 0;
    for (int is1 = 0; is1 < 2; is1++)
        for (int is2 = 0; is2 < 2; is2++) {
            int s1 = SIGMAS[is1];
            int s2 = SIGMAS[is2];

            int factor = 1;
            factor *= arg_s1 ? s1 : 1;
            factor *= arg_s2 ? s2 : 1;

            result += factor * exp((s1 - 1) * hi + (s2 - 1) * hj + (s1 * s2 - 1) * l);
        }

    return result;
}

double E_not_symmetrical(double hi, double hj, double l, bool arg_s1, bool arg_s2) {
    return E_norm_not_symmetrical(hi, hj, l, arg_s1, arg_s2) * exp(hi + hj + l);
}
